### PR TITLE
Api rewrite

### PR DESCRIPTION
**Pull request by synthtech**
This creates a new Anilist api wrapper that could replace the existing `generalAPIcall` and `authAPIcall` functions. The existing wrapper requires the use of many nested callbacks when creating multiple requests leading to code that becomes difficult to read/understand (imo).

I figured it wouldn't be worthwhile to change the existing wrapper and instead wrote a new one in parallel, that does not depend on the existing implementation. It's not a drop-in replacement, so module code would have to be reworked to use it (somewhat similar to the `exportModule` migration.)

A few differences introduced with `anilistAPI`:
- Asynchronous
- Request options are sent in a single argument
- Added authentication as an option
- Added internal api as an option
- Cache backed by localforage
- Rate limiting

It should be feature-complete with the exception of a `queryPacker` replacement, though I plan on doing this at some point. I've added some JSDoc-style documentation that should provide a brief explanation of the new functions.

---
Usage comparison:

Old
```js
generalAPIcall(
  "query($id:Int){Media(id:$id){title {userPreferred} type startDate{year} }}",
  {id: 132473},
  function(data){
    if(!data){
      return
    }
    if(data.data.Media.startDate.year < 2000){
      authAPIcall(
        "mutation($id:Int){SaveMediaListEntry(mediaId:$id,status:PLANNING)}",
          {id: 132473},
          function(data){
            if(!data){
              return
            }
            someCallback(data.data)
          }
      )
    }
  },
  "hohMediaThing" + 132473,10*60*1000,true
)
```
New
```js
const {data, errors} = await anilistAPI("query($id:Int){Media(id:$id){title {userPreferred} type startDate{year} }}", {
  variables: {id: 132473},
  cacheKey: "mediaThing" + 132473,
  duration: 10*60
})
if(errors){
  return
}
if(data.Media.startDate.year < 2000){
  const {data, errors} = await anilistAPI("mutation($id:Int){SaveMediaListEntry(mediaId:$id,status:PLANNING)}", {
    variables: {id: 132473},
    auth: true
  })
  if(errors){
    return
  }
  return data
}
return
```

### CODE REVIEW

**Comment by hohMiyazawa** [comment]
Wow, that's a monumental task.

As is probably evident, the original wrapper was built out with new feature as they were needed, eventually ending up with a less than ideal design.


>    Asynchronous

Good, altough I'm not really agaist nested callbacks.

>    Request options are sent in a single argument

Good

>    Added authentication as an option

Good, although I want to review this closely (e.g. use the wrapper for some modules first) before permitting another way of authentication in modules. No code change required, just don't try replacing ``authAPIcall``white refactoring other modules in the first few weeks

>    Added internal api as an option

Good

>    Cache backed by localforage

Very good

>    Rate limiting

Very good, old system wasn't even usable

File: src/graphql.js [hunk 1]
```diff
@@ -845,5 +845,23 @@ async function anilistAPI(query, queryArgs){
...
+/**
+ * Runs an API cache checkup weekly
+ * @returns {Promise|null}
+ */
+const cacheCheckup = async () => {
+	const check = localStorage.getItem("automail-db-check")
+	if(check){
+		if(NOW() > check){
+			localStorage.setItem("automail-db-check", NOW()+7*24*60*60*1000)
+			return await flushCache();
```
Comment (hohMiyazawa) [inline src/graphql.js]: Why is this needed?

File: src/graphql.js [hunk 2]
```diff
@@ -628,4 +628,258 @@ function authAPIcall(query,variables,callback,cacheKey,timeFresh,useLocalStorage
...
+			apiCache.removeItem(key);
+		}
+	})
+}
+
+/**
+ * Saves data to the API cache
+ * @param {string} key - The key used to store the data.
+ * @param {object} data - The data to store.
+ * @param {number} [duration] - The length of time to store in seconds.
```
Comment (hohMiyazawa) [inline src/graphql.js:732]: Why change the API to seconds instead of milliseconds? My reasoning for picking ``ms`` was that when you write ``*1000``, there's no ambiguity.

**Comment by hohMiyazawa** [comment]
``queryPacker`` reeplacement: This needs some careful design considerations. Query complexity is limited, so it needs a good heuristic for when to pack queries, when to send multiple, and when to send some of the request, relaying partial data back to the module, and waiting out the API rate limiting time to send the rest of the requests.

Opinion: ``queryPacker`` is a low priority project, as the script has more pressing problems.

**Comment by hohMiyazawa** [comment]
Does an API v2 warrant a significant version number bump?

File: src/graphql.js [hunk 3]
```diff
@@ -628,4 +628,258 @@ function authAPIcall(query,variables,callback,cacheKey,timeFresh,useLocalStorage
...
+		expiresAt: duration ? NOW() + duration*1000 : undefined
+	};
+	try{
+		return apiCache.setItem(key, saltedHam);
+	}
+	catch(e){
+		if(e.name === "QuotaExceededError"){
+			console.error("Persistent storage quota exceeded. Attempting to purge expired items.")
+			try{
+				return flushCache();
```
Comment (hohMiyazawa) [inline src/graphql.js:748]: But what if purging expired items is not enough?

One could imagine a more aggressive pruning strategy, based on item size, time, and number of previous cache hits that attempts to remove even more items if the first round of pruning is not enough.

File: src/graphql.js (hunk 1 above)
Comment (synthtech) [reply src/graphql.js]: Requests that have been cached with an expiry date wouldn't get removed until the same call is made again. I was imagining a scenario where some of these expired cache items would continue to accumulate without ever being "flushed", so the solution to that would either be an automated "checkup" that periodically iterates through and removes the expired items or a button in the automail settings that a user could manually "flush" the database.

Was inspired by this bit of code here:
https://github.com/hohMiyazawa/Automail/blob/9ac16a628e3bed9389b32d09e910e3111a863bd3/src/controller.js#L403-L412

Comment (hohMiyazawa) [reply src/graphql.js]: Okay, makes sense. Expiration time can be tweaked later as needed.

File: src/graphql.js (hunk 2 above)
Comment (synthtech) [reply src/graphql.js:732]: I thought it would simplify things since I didn't think that kind of granularity was needed. I'm not against changing it back though.

Comment (hohMiyazawa) [reply src/graphql.js:732]: If there's no strong reason for the change, I prefer keeping it the same as in the v1 api to make conversion of modules less risky.

**Comment by hohMiyazawa** [comment]
Looks good to me.

Merging after a night's sleep.

File: src/graphql.js (hunk 3 above)
Comment (synthtech) [reply src/graphql.js:748]: I mean, at that point I wonder if the user would have enough free disk space for the pruning to even make a difference.

IndexedDB is generally supposed to have a much larger limit than the 5MB that Localstorage has.
https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria#storage_limits

I was previously considering dropping the store entirely as a last resort with [`apiCache.clear()`](https://localforage.github.io/localForage/#data-api-clear)

Comment (hohMiyazawa) [reply src/graphql.js:748]: That sounds better.

**Comment by synthtech** [comment]
>just don't try replacing authAPIcall while refactoring other modules in the first few weeks

sure thing
>Opinion: queryPacker is a low priority project, as the script has more pressing problems.

I agree, i saw it was only used in a few modules so i didn't prioritize it.
>Does an API v2 warrant a significant version number bump?

sure, it feels like it's been on 9.99.x forever lol
>Looks good to me.

👍

File: src/graphql.js (hunk 2 above)
Comment (synthtech) [reply src/graphql.js:732]: ok i will change it back soon

File: src/graphql.js (hunk 3 above)
Comment (synthtech) [reply src/graphql.js:748]: added in f97a77c

**Comment by hohMiyazawa** [comment]
@synthtech 

> just don't try replacing authAPIcall while refactoring other modules in the first few weeks

No longer the case, auth can be freely used now, just with the usual rule of thumb of not using it when it
s not needed.
I see you for instance have a submission stats experiment?

**Comment by synthtech** [comment]
>auth can be freely used now

aight 👍 

>I see you for instance have a submission stats experiment?

yea, i was holding that back since its a new feature. it's basically ready though (needs rebasing), will send a PR later
